lib/kino/table.ex

```diff
@@ -162,19 +153,16 @@ defmodule Kino.Table do
       filters: ctx.assigns.filters
     }
 
-    {:ok, %{columns: columns, data: data, total_rows: total_rows}, state} =
+    {:ok, %{columns: columns, data: {orientation, data}, total_rows: total_rows}, state} =
       ctx.assigns.module.get_data(rows_spec, ctx.assigns.state)
 
-    {columns, data, key_to_string} = stringify_keys(columns, data, ctx.assigns.key_to_string)
+    {columns, key_to_string} = stringify_keys(columns, ctx.assigns.key_to_string)
 
     ctx = assign(ctx, state: state, key_to_string: key_to_string)
 
-    row_based = is_list(data)
-    page_length = if row_based, do: length(data), else: Enum.at(data, 0) |> elem(1) |> length()
-
-    sample_data =
-      if row_based, do: List.first(data), else: Map.values(data) |> Enum.map(&List.first/1)
-
+    row_based = orientation == :row_based
+    page_length = if row_based, do: length(data), else: hd(data) |> length()
+    sample_data = if row_based, do: List.first(data), else: Enum.map(data, &List.first(&1))
```

Comment:
Probably cleaner to do a `case` and return a tuple with `{page_length, sample_data}` :)